components/event-publisher-proxy/pkg/handler/nats/handler_test.go

```diff
@@ -0,0 +1,453 @@
+package nats
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"reflect"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	dynamicfake "k8s.io/client-go/dynamic/fake"
+
+	"github.com/nats-io/nats-server/v2/server"
+	"github.com/nats-io/nats.go"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/kyma-project/kyma/components/event-publisher-proxy/pkg/env"
+	"github.com/kyma-project/kyma/components/event-publisher-proxy/pkg/handler/handlertest"
+	"github.com/kyma-project/kyma/components/event-publisher-proxy/pkg/informers"
+	"github.com/kyma-project/kyma/components/event-publisher-proxy/pkg/legacy-events"
+	"github.com/kyma-project/kyma/components/event-publisher-proxy/pkg/options"
+	"github.com/kyma-project/kyma/components/event-publisher-proxy/pkg/receiver"
+	"github.com/kyma-project/kyma/components/event-publisher-proxy/pkg/sender"
+	"github.com/kyma-project/kyma/components/event-publisher-proxy/pkg/subscribed"
+	testingutils "github.com/kyma-project/kyma/components/event-publisher-proxy/testing"
+	eventingv1alpha1 "github.com/kyma-project/kyma/components/eventing-controller/api/v1alpha1"
+	eventingtesting "github.com/kyma-project/kyma/components/eventing-controller/testing"
+)
+
+type Test struct {
+	logger     *logrus.Logger
+	natsConfig *env.NatsConfig
+	natsServer *server.Server
+}
+
+func (t *Test) setup() {
+	port := testingutils.GeneratePortOrDie()
+	natsPort := testingutils.GeneratePortOrDie()
+	t.natsConfig = newEnvConfig(port, natsPort)
+	t.logger = logrus.New()
+	t.natsServer = eventingtesting.RunNatsServerOnPort(port + 1)
+}
+
+func (t *Test) cleanup() {
+	t.natsServer.Shutdown()
+}
+
+var test = Test{}
+
+func TestMain(m *testing.M) {
+	test.setup()
+	code := m.Run()
+	test.cleanup()
+	os.Exit(code)
+}
+
+func TestNatsHandlerForCloudEvents(t *testing.T) {
+	exec := func(t *testing.T, applicationName, expectedNatsSubject string) {
+		// test environment
+		var (
+			healthEndpoint  = fmt.Sprintf("http://localhost:%d/healthz", test.natsConfig.Port)
+			publishEndpoint = fmt.Sprintf("http://localhost:%d/publish", test.natsConfig.Port)
+		)
+
+		test.logger.Info("TestNatsHandler started")
+
+		// a cancelable context to be used
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		// configure message receiver
+		messageReceiver := receiver.NewHttpMessageReceiver(test.natsConfig.Port)
+		assert.NotNil(t, messageReceiver)
+
+		// create a Nats sender
+		natsUrl := test.natsServer.ClientURL()
+		assert.NotEmpty(t, natsUrl)
+		msgSender := sender.NewNatsMessageSender(ctx, natsUrl, test.logger)
+
+		// configure legacyTransformer
+		appLister := handlertest.NewApplicationListerOrDie(ctx, applicationName)
```

Comment:
Can we scoop informer set up and make it common helpers? It is applicable for HTTP handler as well. `exec` is huge IMO.

Reply:
Done.